File: testjs/fraud.js
var d = require('../libjs/durable');

with (d.ruleset('fraudDetection')) {
    whenAll(span(10), 
            or(m.t.eq('debitCleared'), m.t.eq('creditCleared')), 
    function(c) {
        var debitTotal = 0;
        var creditTotal = 0;
        if (c.s.balance == null) {
            c.s.balance = 0;
            c.s.avgBalance = 0;
            c.s.avgWithdraw = 0;
        } 

        for (var i = 0; i < c.m.length; ++i) {
            if (c.m[i].t === 'debitCleared') {
                debitTotal += c.m[i].amount;
            } else {
                creditTotal += c.m[i].amount;
            }
        }

        c.s.balance = c.s.balance - debitTotal + creditTotal;
        c.s.avgBalance = (c.s.avgBalance * 29 + c.s.balance) / 30;
        c.s.avgWithdraw = (c.s.avgWithdraw * 29 + debitTotal) / 30;
    });

    whenAll(c.first = and(m.t.eq('debitRequest'), 
                          m.amount.gt(c.s.avgWithdraw.mul(3))),
            c.second = and(m.t.eq('debitRequest'),
                           m.amount.gt(c.s.avgWithdraw.mul(3)),
                           m.stamp.gt(c.first.stamp),
                           m.stamp.lt(c.first.stamp.add(90))),
    function(c) {
        console.log('Medium risk ' + c.first.amount + ' ,' + c.second.amount);
    });

    whenAll(c.first = m.t.eq('debitRequest'),
            c.second = and(m.t.eq('debitRequest'),
                           m.amount.gt(c.first.amount),
                           m.stamp.lt(c.first.stamp.add(180))),
            c.third = and(m.t.eq('debitRequest'),
                          m.amount.gt(c.second.amount),
                          m.stamp.lt(c.first.stamp.add(180))),
            s.avgBalance.lt(add(c.first.amount, c.second.amount, c.third.amount).div(0.9)),
    function(c) {
        console.log('High risk ' + c.first.amount + ' ,' + c.second.amount + ' ,' + c.third.amount);
    });

    whenAll(or(m.t.eq('customer'), timeout('customer')), 
    function(c) {
        if (!c.s.cCount) {
            c.s.cCount = 100;
        } else {
            c.s.cCount += 2;
        }

        c.post('fraudDetection', {id: c.s.cCount, sid: 1, t: 'debitCleared', amount: c.s.cCount});
        c.post('fraudDetection', {id: c.s.cCount + 1, sid: 1, t: 'creditCleared', amount: (c.s.cCount - 100) * 2 + 100});
        c.startTimer('customer', 1, 'c' + c.s.cCount);
    });

    whenAll(or(m.t.eq('fraudster'), timeout('fraudster')), 
    function(c) {
        if (!c.s.fCount) {
            c.s.fCount = 1000;
        } else {
            c.s.fCount += 1;
        }

        c.post('fraudDetection', {id: c.s.fCount, sid: 1, t: 'debitRequest', amount: c.s.fCount - 800, 'stamp': Date.now() / 1000});
        c.startTimer('fraudster', 2, 'f' + c.s.cCount);
    });

    whenStart(function (host) {
        host.post('fraudDetection', {id: 'c_1', sid: 1, t: 'customer'});
        host.post('fraudDetection', {id: 'f_1', sid: 1, t: 'fraudster'});
    });
}

d.runAll();

File: testrb/fraud.rb
require "durable"

Durable.ruleset :fraud_detection do
  when_all span(10), (m.t == "debit_cleared") | (m.t == "credit_cleared") do
    debit_total = 0
    credit_total = 0
    if not s.balance
      s.balance = 0
      s.avg_balance = 0
      s.avg_withdraw = 0
    end    

    for tx in m do
      if tx.t == "debit_cleared"
        debit_total += tx.amount
      else
        credit_total += tx.amount
      end
    end

    s.balance = s.balance - debit_total + credit_total
    s.avg_balance = (s.avg_balance * 29 + s.balance) / 30
    s.avg_withdraw = (s.avg_withdraw * 29 + debit_total) / 30
  end

  when_all c.first = (m.t == "debit_request") & 
                     (m.amount > s.avg_withdraw * 3),
           c.second = (m.t == "debit_request") & 
                      (m.amount > s.avg_withdraw * 3) & 
                      (m.stamp > first.stamp) &
                      (m.stamp < first.stamp + 90) do
    puts "Medium risk #{first.amount}, #{second.amount}"
  end

  when_all c.first = m.t == "debit_request",
           c.second = (m.t == "debit_request") &
                      (m.amount > first.amount) & 
                      (m.stamp < first.stamp + 180),
           c.third = (m.t == "debit_request") & 
                     (m.amount > second.amount) & 
                     (m.stamp < first.stamp + 180),
           s.avg_balance < (first.amount + second.amount + third.amount) / 0.9 do
    puts "High risk #{first.amount}, #{second.amount}, #{third.amount}"
  end

  when_all timeout("customer") | (m.t == "customer") do
    if not s.c_count
        s.c_count = 100
    else
        s.c_count += 2
    end

    post "fraud_detection", {:id => s.c_count, :sid => 1, :t => "debit_cleared", :amount => s.c_count}
    post "fraud_detection", {:id => s.c_count + 1, :sid => 1, :t => "credit_cleared", :amount => (s.c_count - 100) * 2 + 100}
    start_timer "customer", 1
  end

  when_all timeout("fraudster") | (m.t == "fraudster") do
    if not s.f_count
        s.f_count = 1000
    else
        s.f_count += 1
    end

    post "fraud_detection", {:id => s.f_count, :sid => 1, :t => "debit_request", :amount => s.f_count - 800, :stamp => Time.now.to_i}
    start_timer "fraudster", 2
  end

  when_start do
    start_timer "fraud_detection", {:id => "c_1", :sid => 1, :t => "customer"}
    start_timer "fraud_detection", {:id => "f_1", :sid => 1, :t => "fraudster"}
  end
end

Durable.run_all
